assume/units/dst_components.py

```diff
@@ -2,782 +2,1592 @@
+
+        self.max_capacity = max_capacity
+        self.min_capacity = min_capacity
+        self.time_steps = time_steps
+        self.max_power_charge = (
+            max_capacity if max_power_charge is None else max_power_charge
+        )
+        self.max_power_discharge = (
+            max_capacity if max_power_discharge is None else max_power_discharge
+        )
+        self.efficiency_charge = efficiency_charge
+        self.efficiency_discharge = efficiency_discharge
+        self.initial_soc = initial_soc
+        self.ramp_up = ramp_up
+        self.ramp_down = ramp_down
+        self.storage_loss_rate = storage_loss_rate
+        self.kwargs = kwargs
+
+    def add_to_model(
+        self, model: pyo.ConcreteModel, model_part: pyo.Block
+    ) -> pyo.Block:
         """
-        Calculates the operating cost of the electrolyser based on the electricity price.
+        Creates and returns a Pyomo Block for the storage component.
+
+        Parameters
+        ----------
+        model : pyo.ConcreteModel
+            A Pyomo ConcreteModel object representing the optimization model.
 
+        Returns
+        -------
+        pyo.Block
+            A Pyomo block representing the storage system with variables and constraints.
         """
-        return (
-            b.electrolyser_operating_cost[t]
-            == b.power_in[t] * model.electricity_price[t]
+
+        # Define parameters
+        model_part.max_capacity = pyo.Param(initialize=self.max_capacity)
+        model_part.min_capacity = pyo.Param(initialize=self.min_capacity)
+        model_part.max_power_charge = pyo.Param(initialize=self.max_power_charge)
+        model_part.max_power_discharge = pyo.Param(initialize=self.max_power_discharge)
+        model_part.efficiency_charge = pyo.Param(initialize=self.efficiency_charge)
+        model_part.efficiency_discharge = pyo.Param(
+            initialize=self.efficiency_discharge
+        )
+        model_part.initial_soc = pyo.Param(
+            initialize=self.initial_soc * self.max_capacity
+        )
+        model_part.ramp_up = pyo.Param(initialize=self.ramp_up)
+        model_part.ramp_down = pyo.Param(initialize=self.ramp_down)
+        model_part.storage_loss_rate = pyo.Param(initialize=self.storage_loss_rate)
+
+        # Define variables
+        model_part.soc = pyo.Var(
+            self.time_steps,
+            within=pyo.NonNegativeReals,
+            bounds=(self.min_capacity, self.max_capacity),
+            doc="State of Charge at each time step",
+        )
+        model_part.charge = pyo.Var(
+            self.time_steps,
+            within=pyo.NonNegativeReals,
+            bounds=(0, self.max_power_charge),
+            doc="Charging power at each time step",
+        )
+        model_part.discharge = pyo.Var(
+            self.time_steps,
+            within=pyo.NonNegativeReals,
+            bounds=(0, self.max_power_discharge),
+            doc="Discharging power at each time step",
```

Comment:
if everyhting is configured correctly, namely the efficiencies and prices, this won't be happening. but if something is configured wrong, like charging is and discharging loses no energy, in that case it can happen. But I don't see it worth the transition to a MILP just to fix someone's bad setup :-D